File: src/turtle_demo_controller/turtle_demo_controller/turtle_plant.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose

class TurtlePlant(Node):
    def __init__(self):
        super().__init__('turtle_plant')
        self.get_logger().info("Turtle Plant Node Started")
        
        # Subscribers (터틀심으로부터 데이터 수신)
        self.turtle1_pose_sub = self.create_subscription(
            Pose, 
            "/turtle1/pose", 
            self.turtle1_pose_handler, 
            10
        )
        self.turtle2_pose_sub = self.create_subscription(
            Pose, 
            "/turtle2/pose", 
            self.turtle2_pose_handler, 
            10
        )
        
        # Publishers (상태 정보 전송)
        self.turtle1_state_pub = self.create_publisher(
            Pose,
            "/turtle1/state",
            10
        )
        self.turtle2_state_pub = self.create_publisher(
            Pose,
            "/turtle2/state",
            10
        )
        
        # 제어 명령 수신 및 전달
        self.control_command_sub = self.create_subscription(
            Twist,
            "/turtle2/control_command",
            self.control_command_handler,
            10
        )
        
        # Publisher for turtle2 control
        self.turtle2_cmd_pub = self.create_publisher(
            Twist,
            "/turtle2/cmd_vel",
            10
        )

    def turtle1_pose_handler(self, msg: Pose):
        """터틀1의 위치 정보 처리"""
        self.turtle1_state_pub.publish(msg)
        self.get_logger().debug(f"Published turtle1 state: x={msg.x:.2f}, y={msg.y:.2f}")

    def turtle2_pose_handler(self, msg: Pose):
        """터틀2의 위치 정보 처리"""
        self.turtle2_state_pub.publish(msg)
        self.get_logger().debug(f"Published turtle2 state: x={msg.x:.2f}, y={msg.y:.2f}")

    def control_command_handler(self, msg: Twist):
        """제어 명령을 터틀2에게 전달"""
        self.turtle2_cmd_pub.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    plant_node = TurtlePlant()
    rclpy.spin(plant_node)
    plant_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main() 

File: src/enc_turtle_cpp/src/turtle_plant.cpp
#include <memory>
#include <string>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"

class TurtlePlant : public rclcpp::Node
{
public:
  TurtlePlant()
  : Node("turtle_plant")
  {
    RCLCPP_INFO(this->get_logger(), "Turtle Plant Node Started");
    
    // Subscribers
    turtle1_pose_sub_ = this->create_subscription<turtlesim::msg::Pose>(
      "/turtle1/pose", 10,
      std::bind(&TurtlePlant::turtle1_pose_callback, this, std::placeholders::_1));
      
    turtle2_pose_sub_ = this->create_subscription<turtlesim::msg::Pose>(
      "/turtle2/pose", 10,
      std::bind(&TurtlePlant::turtle2_pose_callback, this, std::placeholders::_1));
      
    // Publishers
    turtle1_state_pub_ = this->create_publisher<turtlesim::msg::Pose>(
      "/turtle1/state", 10);
      
    turtle2_state_pub_ = this->create_publisher<turtlesim::msg::Pose>(
      "/turtle2/state", 10);
      
    turtle2_cmd_pub_ = this->create_publisher<geometry_msgs::msg::Twist>(
      "/turtle2/cmd_vel", 10);
      
    // Control command subscriber
    control_command_sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
      "/turtle2/control_command", 10,
      std::bind(&TurtlePlant::control_command_callback, this, std::placeholders::_1));
  }

private:
  void turtle1_pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
  {
    // 상태 정보 전송
    turtle1_state_pub_->publish(*msg);
    RCLCPP_DEBUG(this->get_logger(), 
      "Published turtle1 state: x=%.2f, y=%.2f", msg->x, msg->y);
  }

  void turtle2_pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
  {
    turtle2_state_pub_->publish(*msg);
    RCLCPP_DEBUG(this->get_logger(), 
      "Published turtle2 state: x=%.2f, y=%.2f", msg->x, msg->y);
  }

  void control_command_callback(const geometry_msgs::msg::Twist::SharedPtr msg)
  {
    turtle2_cmd_pub_->publish(*msg);
  }

  // ROS2 subscribers
  rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr turtle1_pose_sub_;
  rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr turtle2_pose_sub_;
  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr control_command_sub_;
  
  // ROS2 publishers
  rclcpp::Publisher<turtlesim::msg::Pose>::SharedPtr turtle1_state_pub_;
  rclcpp::Publisher<turtlesim::msg::Pose>::SharedPtr turtle2_state_pub_;
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr turtle2_cmd_pub_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TurtlePlant>());
  rclcpp::shutdown();
  return 0;
}
